### PR TITLE
native interface: access specific parts of request

### PR DESCRIPTION
**Pull request by vividsnow**
Hello

This patch gives access to specific parts of feersum connection/request: ```method uri path query content_length remote_address remote_port headers input```. Thus gives performance boost in some scenarios. 

Serving native handler via unix-socket behind nginx: 
```perl
sub ($r) {
    my ($method, $path) = ($r->method, $r->path);
    $r->send_response(200, [qw'content-type text/plain'], ["$method:$path"]);
}
```
is ~15% faster than psgi handler
```perl
sub ($env) {
    my ($method, $path) = @$env{uc'request_method', uc'path_info'};
    [200, [qw'content-type text/plain'], ["$method:$path"]]
}
```

p.s. there is no performance penalty comparing to current version
p.p.s. tested under linux, gcc 10, perl-5.34.0, perl-5.8.9

### CODE REVIEW

**Comment by audreyt** [comment]
Hi! Thank you for the contribution. Sorry for the late reply, but I currently do not have the capacity to review & merge this pull request.

If @stash can chime in that would be great, otherwise we may have to wait until someone volunteers time to assume maintainership.

**Comment by vividsnow** [comment]
Hi! I'd like to help maintain Feersum if nobody is currently doing so, and move it a bit forward by adding features such as accept4, tcp defer and keepalive support. I've already [worked](https://github.com/vividsnow/Feersum/tree/keepalive-support) on these features and have used them in production. [my cpan](https://metacpan.org/author/egor)

cheers

**Comment by stash** [comment]
Hi @vividsnow - I'm going to _try_ check this PR over, but fair warning that I'm incredibly rusty at Perl. I would love for all of our work here to be maintained... @audreyt if you have a moment can you DM/email me your thoughts? Thank you both for carrying the torch.

Can you please reference an official source for `ppport.h`? 14kLoC is too large to review, so I'd like to diff against that.

File: Feersum.xs [hunk 1]
```diff
@@ -953,7 +977,7 @@ try_parse_http(struct feer_conn *c, size_t last_read)
 
     return phr_parse_request(SvPVX(c->rbuf), SvCUR(c->rbuf),
         &req->method, &req->method_len,
-        &req->path, &req->path_len, &req->minor_version,
+        &req->uri, &req->uri_len, &req->minor_version,
```
Comment (stash) [inline Feersum.xs:980]: May wish to guard that `uri_len` isn't zero

File: Feersum.xs [hunk 2]
```diff
@@ -1461,6 +1484,107 @@ needs_decode:
...
+{
+    return (r->minor_version == 1) ? psgi_serv11 : psgi_serv10;
+}
+
+INLINE_UNLESS_DEBUG static void
+feersum_set_path_and_query(pTHX_ struct feer_req *r)
+{
+    const char *qpos = r->uri;
+    while (*qpos != '?' && qpos < r->uri + r->uri_len) qpos++;
+    if (*qpos == '?') {
```
Comment (stash) [inline Feersum.xs:1547]: If `uri_len` is zero, then these aren't necessarily valid pointer de-references. So, probably you could assert `uri_len > 0` here too.

File: Feersum.xs [hunk 3]
```diff
@@ -1673,6 +1743,74 @@ feersum_env(pTHX_ struct feer_conn *c)
     return e;
 }
 
+#define NORM_HEADER(_str) \
+char *k = kbuf;\
+for (j = 0; j < hdr->name_len; j++) { char n = hdr->name[j]; *k++ = _str; }
```
Comment (stash) [inline Feersum.xs:1748]: Normally, I'd complain "`_str` is being evaluated multiple times!!" but, this is being clever ;)

File: Feersum.xs [hunk 4]
```diff
@@ -1673,6 +1743,74 @@ feersum_env(pTHX_ struct feer_conn *c)
...
+
+#define COPY_HEADER(_norm) \
+for (i = 0; i < r->num_headers; i++) {\
+    struct phr_header *hdr = &(r->headers[i]);\
+    if (unlikely(hdr->name == NULL && val != NULL)) {\
+        sv_catpvn(*val, hdr->value, hdr->value_len);\
+        continue;\
+    }\
+    _norm\
```
Comment (stash) [inline Feersum.xs:1757]: You're not using `NORM_HEADER` outside of `COPY_HEADER`, so consider in-lining it here to simplify.

**Review by stash — COMMENTED** [review]
Made a few suggestions, which you're free to use as you wish.

**Comment by stash** [comment]
I'll wait until 2024-07-30T19:00:00-08:00 (7PM, Tues, June 30th, PDT, about a week from now) to hear back from @audreyt, who's been the main maintainer for years, or any of the other contributors, before taking action on adding you as co-maintainer or transferring maintainership.

If somehow I disappear again, https://www.cpan.org/modules/04pause.html outlines the steps to take over, but we'll try the direct approach first ;)

**Comment by stash** [comment]
Oh maybe @audreyt already has it: https://metacpan.org/dist/Feersum

File: Feersum.xs (hunk 1 above)
Comment (vividsnow) [reply Feersum.xs:980]: phr_parse_request sets it to zero before parsing start: https://github.com/h2o/picohttpparser/blob/f8d0513f1a7a111f2597d643b073935a8afaf9e5/picohttpparser.c#L411

File: Feersum.xs (hunk 2 above)
Comment (vividsnow) [reply Feersum.xs:1547]: According to spec: https://datatracker.ietf.org/doc/html/rfc7230#section-5.3.1 empty path must be specified by "/", so phr_parse_request will fail earlier if length of path component is 0: https://github.com/h2o/picohttpparser/blob/f8d0513f1a7a111f2597d643b073935a8afaf9e5/picohttpparser.c#L381

**Comment by vividsnow** [comment]
@stash Hi, thanks for the review. I'll check your points.

ppport.h was rendered by specific version of Devel::PPPort
https://metacpan.org/release/ATOOMIC/Devel-PPPort-3.62/view/PPPort.pm - I'll update it once again with current Devel::PPPort version (https://metacpan.org/pod/Devel::PPPort, which is 3.68)

**Comment by stash** [comment]
> ppport.h was rendered by specific version of Devel::PPPort https://metacpan.org/release/ATOOMIC/Devel-PPPort-3.62/view/PPPort.pm - I'll update it once again with current Devel::PPPort version (https://metacpan.org/pod/Devel::PPPort, which is 3.68)

Looks like Devel::PPPort 3.72 ships with Perl 5.40.0, and the file checked in matches that. The CPAN one is outdated now. Looks good; just FYI.

**Review by stash — APPROVED** [review]

**Comment by stash** [comment]
Merged. Let's wait to see if anyone else responds before releasing to CPAN. FYI looks like picohttpparser has some recent updates.

**Comment by stash** [comment]
Alright @vividsnow, it's time: invited you to collab on this repo and added you as comaint on PAUSE. Thank you so much for your contributions and willingness to carry this forward. Please email me (address is on CPAN) if you need anything else. I can try do some code review if you'd like, but don't wait for me!

**Comment by vividsnow** [comment]
Thank you for the trust, @stash. I'll do my best

**Pull request merged**